client/src/screens/PrintTestDeckScreen.tsx

```diff
@@ -130,44 +79,105 @@ const TestDeckBallotsMemoized = React.memo(TestDeckBallots)
 const PrintTestDeckScreen = () => {
   const { electionDefinition } = useContext(AppContext)
   const { election, electionHash } = electionDefinition!
+  const [precinctIds, setPrecinctIds] = useState<string[]>([])
+  const [precinctIndex, setPrecinctIndex] = useState<number | undefined>(
+    undefined
+  )
+
   const { precinctId: p = '' } = useParams<PrecinctReportScreenProps>()
   const precinctId = p.trim()
 
-  const [printEnabled, setPrintEnabled] = useState(false)
-
   const pageTitle = 'Test Deck'
-  const precinct =
+  const precinctName =
     precinctId === 'all'
-      ? allPrecincts
-      : getPrecinctById({ election, precinctId })
+      ? 'All'
+      : getPrecinctById({ election, precinctId })?.name
 
   useEffect(() => {
-    setPrintEnabled(false)
-  }, [precinctId])
+    if (precinctId) {
+      console.log('got precinctId', precinctId)
+      const precinctIds =
+        precinctId === 'all'
+          ? election.precincts.map((p) => p.id)
+          : [precinctId]
+      console.log('precinctIds are', precinctIds)
+      setPrecinctIds(precinctIds)
+    } else {
+      setPrecinctIds([])
+      setPrecinctIndex(undefined)
+    }
+  }, [precinctId, election.precincts])
+
+  const startPrint = async () => {
+    if (window.kiosk) {
+      const printers = await window.kiosk.getPrinterInfo()
+      if (printers.some((p) => p.connected)) {
+        setPrecinctIndex(0)
+      } else {
+        window.alert('please connect the printer.')
+      }
+    } else {
+      // plain browser
+      setPrecinctIndex(0)
+    }
+  }
 
-  const onAllRendered = useCallback(() => {
-    setPrintEnabled(true)
-  }, [])
+  const onAllRendered = useCallback(
+    async (pIndex, numBallots) => {
+      if (window.kiosk) {
+        await window.kiosk.print()
+      } else {
+        await window.print()
+      }
 
-  if (precinct) {
+      if (pIndex < precinctIds.length - 1) {
+        // wait 5s per ballot printed
+        // that's how long printing takes in duplex, no reason to get ahead of it.
+        await sleep(numBallots * 5000)
+        setPrecinctIndex(pIndex + 1)
+      } else {
+        await sleep(3000)
+        setPrecinctIndex(undefined)
+      }
+    },
+    [setPrecinctIndex, precinctIds]
+  )
+
+  const currentPrecinct =
+    precinctIndex === undefined
+      ? undefined
+      : getPrecinctById({ election, precinctId: precinctIds[precinctIndex] })
+
+  if (precinctIds.length > 0) {
     return (
       <React.Fragment>
+        {precinctIndex !== undefined && currentPrecinct && (
+          <Modal
+            isOpen
+            centerContent
+            content={
+              <Prose>
+                <Loading as="h2">
+                  Printing Test Ballots for {currentPrecinct.name} (#
+                  {(precinctIndex + 1).toString()} of{' '}
+                  {precinctIds.length.toString()})
+                </Loading>
+              </Prose>
```

Comment:
The `Loading` component wasn't intended to be used this way, and already contains a `Prose` tag. So, instead of refactoring it, let's do this…
```suggestion
              <Loading as="p">
                Printing Test Deck
                {precinctIds.length > 1
                  ? ` (${precinctIndex + 1} of ${precinctIds.length})`
                  : ''}
                : {currentPrecinct.name}
              </Loading>
```